update trusted boot docs and add blog post (#168)
kairos-io/kairos-docs: content/en/blog/trusted-boot.md

---
title: "Unlocking the Mysteries of Trusted Boot: A Deep Dive into Secure System Boot Processes"
date: 2024-04-10
linkTitle: "Unlocking the Mysteries of Trusted Boot: A Deep Dive into Secure System Boot Processes"
description: ""
author: Ettore Di Giacinto ([Twitter](https://twitter.com/mudler_it)) ([GitHub](https://github.com/mudler))
---

In the evolving landscape of cybersecurity, protecting the integrity of computing systems from the moment they power on has become very important. As threats become more sophisticated, understanding and implementing advanced boot security mechanisms like Trusted Boot, Full Disk Encryption (FDE), Secure Boot, and Measured Boot are critical for safeguarding data and ensuring system integrity. This article demystifies these concepts, explores their significance, and examines their implementation in modern computing environments, particularly focusing on the Linux ecosystem and the approaches within the Kairos project.

> > We are facing new challenges as 2024 starts. While AI is pushing innovating technologies to the edge, security becomes even more critical for organizations to protect sensitive data.

Author: It should display as a quote (italics) but well..

One of the primary challenges in securing edge devices is their inherent diversity and distributed nature. Unlike centralized systems, edge devices are spread across various locations, often in unsecured or semi-secured environments, making physical and network security a significant concern. This dispersion also implies a broad attack surface, with each device potentially offering a unique set of vulnerabilities. What we are trying to build here is a new foundation to define how to deploy secure devices which cannot be tampered with, and can keep data secure from malicious actors. We do want to protect the software stack that we run on it . 

Our goal is simple but complex to achieve: No unauthorized access to the software, to the data which is generated by the machine, and finally, we don’t want the machine to execute modified code.

## The Essence of Boot Security and Encryption

![DALL·E](https://github.com/kairos-io/kairos-docs/assets/2420543/ea0a4d89-b037-44a2-a89b-7ad586787201)

Here comes Trusted Boot and SENA to the rescue. Easier to explain with an Alice metaphor.

> > you can read more about Trusted Boot in https://0pointer.de/blog/brave-new-trusted-boot-world.html and about SENA here: https://kairos.io/blog/2023/04/18/kairos-is-now-part-of-the-secure-edge-native-architecture-by-spectro-cloud-and-intel/

Author: ```suggestion
> You can read more about Trusted Boot in the [Lennart blog](https://0pointer.de/blog/brave-new-trusted-boot-world.html) and about SENA in our [blog post](https://kairos.io/blog/2023/04/18/kairos-is-now-part-of-the-secure-edge-native-architecture-by-spectro-cloud-and-intel/)
```

> > you can read more about Trusted Boot in https://0pointer.de/blog/brave-new-trusted-boot-world.html and about SENA here: https://kairos.io/blog/2023/04/18/kairos-is-now-part-of-the-secure-edge-native-architecture-by-spectro-cloud-and-intel/

Reviewer: I'd capitalize the first word here and I would avoid the full URLs. I think those look good on footnotes but can be a bit too much here

Let's imagine Alice has this super cool treasure chest, chock-full of her most prized stuff. She’s big on security, but in a high-tech way – kind of like how your computer needs to keep its data safe. To get this done, your computer uses some smart tricks, things like Trusted Boot, Full Disk Encryption (FDE), Secure Boot, and Measured Boot.

To picture Full Disk Encryption, or FDE, think of it as Alice’s treasure chest having a lock that only opens with a secret code. It’s so clever that if someone breaks into her house and finds the chest, they still can’t get inside without that special code. FDE is like that for your computer – it scrambles all your data so that only someone with the right password can read it.

Next, there's Secure Boot. Imagine if Alice had this smart doorbell. When someone presses it, the doorbell checks if they're on the guest list. If they're not, no entry. Secure Boot is your computer's version of this smart doorbell. It checks every piece of software when your computer starts up and only lets the good, trusted stuff in. It keeps out anything that looks fishy.

> And then, we have Measured Boot. Picture this as Alice keeping a super detailed logbook. She writes down everything that goes on in her house – who comes in, what they do, you name it. If anything odd pops up, she’s on it. Measured Boot does something similar for your computer. It watches over the startup process, takes notes, and makes sure everything's running just like it should, no funny business.

Reviewer: Also I don't know about the log book analogy. I might not have a great suggestion to give here but I feel like the trusted boot is more like a bouncer not letting just anyone in, and measured boot like a list of people who can actually come in and who doesn't ... anyways, don't need to change it but I feel it's a bit more one to one

Putting all these together, you’ve got Trusted Boot, which is like Alice's ultimate home security system – the secret-code lock, the smart doorbell, and the detailed logbook, all working together. We’re going to unpack how these nifty tools keep your computer as secure as Alice’s treasure chest, making sure your digital world is safe every time you power up.

## Challenges and Solutions in Implementing FDE on Linux

![DALL·E](https://github.com/kairos-io/kairos-docs/assets/2420543/aee89c19-a5d1-4b75-8b05-bbeef4e2b47e)

> While FDE is a standard feature in various operating systems, its implementation in Linux presents unique challenges. The initial boot stage, including the kernel and initial file system setup, cannot be fully encrypted, leaving a portion of the system exposed. However, through careful measures, this unencrypted segment can be protected against tampering, thus securing the system's boot process and the encrypted data it accesses.

Reviewer: As a reader I would like to know why it isn't possible to fully encrypt. Doesn't need to be long but at least one phrase. Again, not critical, feel free to ignore

Author: good point! will address that in my next changeset

> The Kairos project introduces "Unified Kernel Images" (UKI) to enhance boot security. UKI files are single, fat binaries that contain the OS and necessary boot components in a single, verified file. This approach simplifies the boot process, allowing for the entire system to be booted securely and efficiently. By leveraging EFI files that can be signed and verified through Secure Boot and measured as a single entity, UKI files represent a significant advancement in boot security.

Reviewer: How about adding one more sentence or a new paragraph explaining that it also comes with challenges? like the relationship with size and memory, or so

> Central to the Trusted Boot mechanism is the Trusted Platform Module (TPM), a dedicated hardware(also emulated by firmware) component designed for secure cryptographic operations. TPM chips play a critical role in storing encryption keys and measurements securely, enabling the system to verify boot integrity and encrypt user data effectively.

Reviewer: ```suggestion
Central to the Trusted Boot mechanism is the Trusted Platform Module (TPM), a dedicated hardware (also emulated by firmware) component designed for secure cryptographic operations. TPM chips play a critical role in storing encryption keys and measurements securely, enabling the system to verify boot integrity and encrypt user data effectively.
```

## From Grub to systemd-boot

![DALL·E 2024-01-18 18 22 47 - Create a vintage-style digital artwork that visualizes the concept of modernizing technology infrastructure  The central focus should be a retro-futur](https://github.com/kairos-io/kairos-docs/assets/2420543/df3bce6f-4e88-4edd-83b2-ba2d333d9e6f)

> Trusted Boot is deeply integrated in systemd and currently under active development. Systemd is renowned for being more than a mere init system. It includes a range of tools like systemd-measure and ukify, which, though compatible with various bootloaders, synergize exceptionally well with systemd-boot. This makes a compelling case for choosing systemd, and particularly systemd-boot.

Reviewer: I think this paragraph could be improved if the line of thought went into. There are other bootloaders, they work well on linux, but systemd-boot is prefered for these reasons. No need to mention that systemd is more than an init system.

> Why this makes sense: We're now in an era where outdated hardware is not our concern. Modern devices don't require old drivers, so we have the opportunity to streamline our systems. This translates to a principle of 'less is more' in our code and in our stack - less software in our stack means enhanced security. We aim for a leaner runtime to run our software, minimizing potential security vulnerabilities. In addition, it also makes sense to use tools that adhere closely to standard specifications.

Reviewer 2: I feel this sections clarification. Why is outdated hardware not a concern anymore? What devices required old drivers (as opposed to "modern" ones)? What kind of drivers are we talking about? I understand this ties to the grub section below but was grub really about "drivers"?

> Considering the transition to modern hardware, especially with widespread EFI support, systemd-stub or systemd-boot is a strong contender to replace GRUB. GRUB has been a staple in the Linux community for years, evolving from versions 0.x to 2.x(and if you are old like me, you have witnessed almost all of the versions!). However, it's worth considering the extensive patches applied by distributions and its expansive codebase. While its extensive support and scripting capabilities were once advantageous, they now detract from its suitability as a lightweight option.

Reviewer: ```suggestion
Considering the transition to modern hardware, especially with widespread EFI support, systemd-stub or systemd-boot is a strong contender to replace GRUB. GRUB has been a staple in the Linux community for years, evolving from versions 0.x to 2.x, which if you are seasoned like me, you have witnessed almost all of them! However, it's worth considering the extensive patches applied by distributions and its expansive codebase. While its extensive support and scripting capabilities were once advantageous, they now detract from its suitability as a lightweight option.
```

The goal is clear: we aim to minimize complexity in building a secure, efficient stack, reusing as much components well trusted and established by the community. By reducing the number of components, we strive for a more secure and streamlined system.

## Booting in Linux

When a Linux system starts, it goes through a multi-stage process to load the operating system. This process can be broken down roughly into the following steps:

1. Firmware Initialization: The system's firmware, which could be BIOS or UEFI, performs initial hardware checks and configurations. It then searches for a bootloader, which is typically located in the boot partition or the Master Boot Record (MBR) of the storage device.

> 1. Bootloader Execution: The bootloader, such as GRUB, is a small program responsible for loading the main part of the operating system. Its primary task is to find the Linux kernel, usually located in the same partition, and load it into memory. The bootloader might present a menu or configuration options to choose different kernels or operating systems if multiple are installed.

Reviewer: I would remove the "small program" to avoid a counter argument to what was said before

> 1. Kernel Loading and Initialization: The kernel is the core of the Linux operating system. Once loaded, it initializes the system's hardware and sets up essential services. During this phase, the kernel decompresses and loads an initrd (initial ramdisk). Role of Initrd (Initial Ramdisk): The initrd is a temporary root file system loaded into memory. It includes essential tools and scripts needed to mount the real root file system. In modern Linux systems, initrd is often replaced by initramfs, which serves a similar purpose.

Reviewer 2: Removing some duplication

```suggestion
1. Kernel Loading and Initialization: The kernel is the core of the Linux operating system. Once loaded, it initializes the system's hardware and sets up essential services. During this phase, the kernel decompresses and loads an initrd (initial ramdisk). The initrd is a temporary root file system loaded into memory. It includes essential tools and scripts needed to mount the real root file system. In modern Linux systems, initrd is often replaced by initramfs, which serves a similar purpose.
```

> 1. File System Setup and Transition: Inside initrd, scripts are executed to prepare the actual root file system. This includes tasks like decrypting encrypted partitions if present. After preparations are complete, the system makes a transition, referred to as "pivoting", to the actual root file system.

Reviewer: ```suggestion
1. File System Setup and Transition: Inside initrd, scripts are executed to prepare the actual root file system. This includes tasks like decrypting and encrypted partitions if present. After preparations are complete, the system makes a transition, referred to as "pivoting", to the actual root file system.
```

1. Root File System and User Data: The root partition, now mounted as the root file system, contains all the user data, system configurations, binaries, drivers, and other necessary components for the operating system to function. This is where the system runs from once the boot process is complete.

Throughout these stages, the components—firmware, bootloader, kernel, initrd, and the root file system—work in a sequence to load the operating system successfully. 

> This chain is the one we in the Linux and tech industry we are very much familiar with. Several designs and strategies have been developed to secure this stack over time, for instance Verified kernels and drivers, or Secure Boot to sign the bootloader and be able to trust its authenticity. There are various bootloader implementations as well, tied to more secure HW (like TPM devices).

Reviewer: ```suggestion
This chain is the one we in the Linux and tech industry are very much familiar with. Several designs and strategies have been developed to secure this stack over time, for instance Verified kernels and drivers, or Secure Boot to sign the bootloader and be able to trust its authenticity. There are various bootloader implementations as well, tied to more secure HW (like TPM devices).
```

## Booting with a TPM-equipped Hardware

When a device equipped with TPM, and a sufficiently modern hardware boots, several steps happen sequentially.

Many hardware platforms use a Core Root of Trust for Measurements (CRTM), which is the very first thing that boots, even before the firmware. The CRTM gets a hash of the firmware and sends it to the TPM chip, which measures the running software and is a requirement for trusted boot. The TPM chip then loads the firmware/BIOS. 

The TPM takes note of the measurement and stores the hash in a bank of multiple platform configuration registers (PCRs). In order to store measurements, the TPM chip extends the banks from the previous values, as these changes to the stack are easy to recognize during the boot.

![CRTM trusted boot](https://github.com/kairos-io/kairos-docs/assets/2420543/5a0244c3-ff58-4ea0-83b3-b391ccac0b91)

Next, the BIOS or firmware measures the subsequent stage (bootloader or UKI) and sends it to the TPM, then loads the UKI and continues booting.

> Once the TPM is asked to release an encryption key to unseal the full disk encryption, it will check if the measurements it has are valid. If a bootloader is present, it will have already measured the UKI files, and measurement also happens when initrd starts and when the running system is ready to check eventual manipulations (e.g., kernel boot command line). The process is also bound to the secure boot signatures, so any manipulation of the UKI files would be allowed only by the key holders.

Reviewer 2: What are "eventual manipulations"? Eventual has a time sense to it (something that will guaranteed to happen as some point in time). I think here we are talking about "booting options" or something like that.

> If a malicious attacker tries to load custom software or modify an image in any way, they must extend the PCR value and align it with the value it would have had following an expected boot. The cryptographic robustness of the hash algorithm makes achieving this computationally expensive, thereby helping to support security. 

Reviewer 2: If an attacker had a way to extend the PCR values, they could make the system trust any modified image, no? That's not computationally expensive.
It would be computationally expensive (or impossible) to modify an image so that it still matches the measurement (if at all possible). Is this correct?

Author: you can always extend the PCR values, however, you would have to expand the PCR values in a way that matches the same exact measurements (that's why it's computationally expensive). It's not only a matter of the image, but it's in a more generic sense

> If the firmware or bootloader have been tampered with, the hash values stored in the PCR won’t match expected measurements, and the system will flag an alert and prevent boot from happening — intentionally bricking the host device as a failsafe. 

Reviewer 2: Bricking would only happen if the firmware was tampered with right? I mean, the bootloader problem is recoverable by re-installing everything.

Author: good point, let me clarify that

## Kairos and Trusted Boot

> ![standardlinux](https://github.com/kairos-io/kairos-docs/assets/2420543/df635d48-e9ff-4759-98d0-3cdaa87ee8b5)

Reviewer 2: The boot partition in this image says "unencrypted" . Is it possible to update the image to say "unencrypted but measured" instead? Just to make sure it doesn't show up as a "hole" in security.

Author: that image is just for standard linux systems, so it is not measured at all

Author: this one (which is the one refered to UKI/EFI) shows measured/verified: https://github-production-user-asset-6210df.s3.amazonaws.com/2420543/315326276-19b61dd1-b33f-4cfe-9d0a-b60198b0cfd5.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240417%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240417T082545Z&X-Amz-Expires=300&X-Amz-Signature=ecab4a7fd4f6ec1a8b2c02134d513e1c164ac910753482dfda029ae5020b066a&X-Amz-SignedHeaders=host&actor_id=2420543&key_id=0&repo_id=632358571

During the boot process, cryptographic assessments are made using the Trusted Platform Module (TPM) to confirm the system's integrity. Drive decryption during boot is contingent upon consistent signing by the same authority and accurate boot measurement matching. System upgrades involve updating the boot measurement to enable access to the user data partitions.

![bootingkeys](https://github.com/kairos-io/kairos-docs/assets/2420543/49ad3793-b53e-40e9-96da-58cec5d4df4c)

## Security consideration

The existing framework guarantees that the initial phase (UKI file) remains immutable, with only this phase capable of decrypting the drive's encrypted data. Indeed there is no pivoting into another system like in the traditional Linux boot process. This design choice ensures that the system remains a single component, with the UKI file acting as the sole point of entry for the system.

> Shifting to a different image introduces potential security vulnerabilities and constraints:

Reviewer 2: I would change this to a "hypothetical" phrasing to avoid people reading this as if it's the case with Kairos:

```suggestion
Shifting to a different image would have introduced potential security vulnerabilities and constraints:
```

the items below need to be adjusted then.
